pkg/roachprod/roachprod.go

```diff
@@ -740,7 +748,77 @@ func Start(
 	if err != nil {
 		return err
 	}
-	return c.Start(ctx, l, startOpts)
+	if err = c.Start(ctx, l, startOpts); err != nil {
+		return err
+	}
+	updatePrometheusTargets(ctx, l, c)
+	return nil
+}
+
+// UpdateTargets updates prometheus target configurations for a cluster.
+func UpdateTargets(
+	ctx context.Context,
+	l *logger.Logger,
+	clusterName string,
+	clusterSettingsOpts ...install.ClusterSettingOption,
+) error {
+	if err := LoadClusters(); err != nil {
+		return err
+	}
+	c, err := newCluster(l, clusterName, clusterSettingsOpts...)
+	if err != nil {
+		return err
+	}
+	updatePrometheusTargets(ctx, l, c)
+	return nil
+}
+
+// updatePrometheusTargets updates the prometheus instance cluster config. Any error is logged and ignored.
+func updatePrometheusTargets(ctx context.Context, l *logger.Logger, c *install.SyncedCluster) {
+	// nodeIPPortInfo is used for sending the node IP information into a channel
+	type nodeIPPortInfo struct {
+		nodeIPPort string
+		nodeIndex  int
+	}
+	// noIpInfoChannel channel is used so that DiscoverService can be executed in parallel
+	noIpInfoChannel := make(chan *nodeIPPortInfo)
+	var wg sync.WaitGroup
+	for i, node := range c.VMs {
+		if node.Provider == gce.ProviderName {
+			wg.Add(1)
+			go func(index int, v vm.VM) {
+				defer wg.Done()
+				// only gce is supported for prometheus
+				desc, err := c.DiscoverService(ctx, install.Node(index+1), "", install.ServiceTypeUI, 0)
+				if err != nil {
+					l.Errorf("error getting the port for node %d: %v", index+1, err)
+					return
+				}
+				nodeInfo := fmt.Sprintf("%s:%d", v.PublicIP, desc.Port)
+				l.Printf("node information obtained for node index %d: %s", index, nodeInfo)
+				noIpInfoChannel <- &nodeIPPortInfo{nodeInfo, index}
+			}(i, node)
+		}
+	}
+	go func() {
+		wg.Wait()
+		close(noIpInfoChannel)
+	}()
+	nodeIPPorts := make([]string, len(c.Nodes))
+	for {
+		info, ok := <-noIpInfoChannel
+		if !ok {
+			break
+		}
+		nodeIPPorts[info.nodeIndex] = info.nodeIPPort
```

Comment:
We're essentially blocking here until all (ip, port) tuples have been "discovered"; i.e., the parallelism doesn't extend to `UpdatePrometheusTargets`, which is a very reasonable trade-off. However, why have all this extra machinery wrt `noIpInfoChannel` and a goroutine for `wg.Wait`? Why not just populate `nodeIPPorts` directly in the outermost goroutine(s) and just wait for it in the current goroutine?

Reply:
`UpdatePrometheusTargets` needs the entire yaml file content, so, we need to make just 1 API call. I used channel just as a precaution as we will be updating `nodeIPPorts` from different go routiness concurrently. But, as they are different indixes, it should be fine.